### PR TITLE
Final P API update.

### PR DESCRIPTION
**Pull request by atsushieno**
API 27 package is updated and it introduced some new API bits too.

### CODE REVIEW

**Comment by jonpryor** [comment]
DP3 hasn't been released yet: https://developer.android.com/preview/overview

APIs are not yet stable.

**Comment by jonpryor** [comment]
Build fails because `platforms/android-28/android.jar` doesn't exist.

...presumably because *it isn't stable yet*.

**Comment by atsushieno** [comment]
The problem is that xamarin-android does not really use "platforms" listed in build-tools/android-toolchains. It uses them from mono/sdks !

I spent not a little time to figure out that.

**Comment by jonpryor** [comment]
Build fails because `platform-28_r04` isn't downloaded, because that download logic has been moved into mono.

This PR needs to include a bump to: https://github.com/mono/mono/commit/6f9bef3bdb83d53578417ea7039081da4e7c9a3d

**Comment by jonpryor** [comment]
[Build is failing](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3321/) because of:

```
The "ResolveSdks" task failed unexpectedly. [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
System.ArgumentNullException: Value cannot be null. [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
Parameter name: element [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at System.Xml.Linq.XElement.op_Explicit (System.Xml.Linq.XElement element) [0x0000e] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Xml.Linq/System/Xml/Linq/XLinq.cs:4525  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at Xamarin.Android.Tools.AndroidVersion.Load (System.Xml.Linq.XDocument doc) [0x00031] in <55cb12b88030401dac208033d851be89>:0  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at Xamarin.Android.Tools.AndroidVersion.Load (System.String uri) [0x00008] in <55cb12b88030401dac208033d851be89>:0  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at Xamarin.Android.Tools.AndroidVersions+<>c.<.ctor>b__16_1 (System.String file) [0x00000] in <55cb12b88030401dac208033d851be89>:0  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at System.Linq.Enumerable+SelectEnumerableIterator`2[TSource,TResult].MoveNext () [0x00036] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/Select.cs:131  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at Xamarin.Android.Tools.AndroidVersions.LoadVersions (System.Collections.Generic.IEnumerable`1[T] versions) [0x00093] in <55cb12b88030401dac208033d851be89>:0  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at Xamarin.Android.Tools.AndroidVersions..ctor (System.Collections.Generic.IEnumerable`1[T] frameworkDirectories) [0x00130] in <55cb12b88030401dac208033d851be89>:0  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at Xamarin.Android.Tasks.MonoAndroidHelper.RefreshSupportedVersions (System.String[] referenceAssemblyPaths) [0x00001] in <0da745fece5743b988534327aec2f476>:0  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
  at Xamarin.Android.Tasks.ResolveSdks.RunTask () [0x001bd] in <0da745fece5743b988534327aec2f476>:0  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/OpenTK-1.0/OpenTK.csproj]
```

**Comment by jonpryor** [comment]
The `ArgumentNullException` is presumably coming somewhere in: https://github.com/xamarin/xamarin-android-tools/blob/50af063b2fdf0eda752882a4a61129e0e55c746c/src/Xamarin.Android.Tools.AndroidSdk/AndroidVersion.cs#L72-L81

My *guess* is that `bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v9.0/AndroidApiInfo.xml` doesn't exist.

To that end I've restarted the build, so we can see what the build outputs *are*. (Hopefully we won't get another/new PR soon...)

**Comment by jonpryor** [comment]
Theory is *close*...but not quite right; from the debug log:

```
Building target "_GenerateAndroidApiInfo" completely.
Output file "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v9.0/AndroidApiInfo.xml" does not exist.
Using "MakeDir" task from assembly "Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
Task "MakeDir" (TaskId:8)
  Task Parameter:Directories=/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v9.0/ (TaskId:8)
  Creating directory "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v9.0/". (TaskId:8)
Done executing task "MakeDir". (TaskId:8)
Added Item(s): _Line=<AndroidApiInfo>
Added Item(s): _Line=</AndroidApiInfo>
Using "WriteLinesToFile" task from assembly "Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
Task "WriteLinesToFile" (TaskId:9)
  Task Parameter:File=/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v9.0/AndroidApiInfo.xml (TaskId:9)
  Task Parameter:
      Lines=
          <AndroidApiInfo>
          </AndroidApiInfo> (TaskId:9)
  Task Parameter:Overwrite=True (TaskId:9)
Done executing task "WriteLinesToFile". (TaskId:9)
Done building target "_GenerateAndroidApiInfo" in project "Mono.Android.csproj".: (TargetId:23)
```

Notice that *there are no lines*; we are (probably) generating an *empty* file, a'la:

```xml
<AndroidApiInfo>
</AndroidApiInfo>
```

File: src/Mono.Android/Mono.Android.projitems [hunk 1]
```diff
@@ -140,10 +140,10 @@
       <Stable>True</Stable>
     </AndroidApiInfo>
     <AndroidApiInfo Include="v8.1.99">
```
Comment (jonpryor) [inline src/Mono.Android/Mono.Android.projitems]: I believe that *this* is the problem: we didn't change it to `v9.0`.  Consequently, when `BuildEverything.mk` builds framework 9.0, `@(_ApiInfo)` is *empty*:

https://github.com/xamarin/xamarin-android/blob/7bd8b7d7748b0734398abf3d966f242ab7eab249/src/Mono.Android/Mono.Android.targets#L138-L141

**Comment by jonpryor** [comment]
[Current failure](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3324/):

```
  "javac" -source 1.5 -target 1.6 -d "obj/Debug/__CreateTestJarFile-bin" -cp "/Users/builder/android-toolchain/sdk/platforms/android-28/android.jar" "java/com/xamarin/android/Bxc4288.java" "java/com/xamarin/android/Bxc58383.java" "java/com/xamarin/android/CallMethodFromCtor.java" "java/com/xamarin/android/CallNonvirtualBase.java" "java/com/xamarin/android/CallNonvirtualDerived.java" "java/com/xamarin/android/DataHandler.java" "java/com/xamarin/android/DataListener.java" "java/com/xamarin/android/Logger.java" "java/com/xamarin/android/MarkerIgnoringBase.java" "java/com/xamarin/android/MyCanvas.java" "java/com/xamarin/android/Timing.java" "java/com/xamarin/android/Bxc9446.java" "java/com/xamarin/android/Bxc7634.java" "java/com/xamarin/android/Bxc37706Throwable.java" (TaskId:1188)
EXEC : warning : [options] bootstrap class path not set in conjunction with -source 1.5 [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/CodeGen-Binding/Xamarin.Android.McwGen-Tests/Xamarin.Android.McwGen-Tests.csproj]
EXEC : warning : [options] source value 1.5 is obsolete and will be removed in a future release [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/CodeGen-Binding/Xamarin.Android.McwGen-Tests/Xamarin.Android.McwGen-Tests.csproj]
EXEC : warning : [options] To suppress warnings about obsolete options, use -Xlint:-options. [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/CodeGen-Binding/Xamarin.Android.McwGen-Tests/Xamarin.Android.McwGen-Tests.csproj]
  java/com/xamarin/android/MyCanvas.java:12: error: method does not override or implement a method from a supertype (TaskId:1188)
    @Override (TaskId:1188)
    ^ (TaskId:1188)
  java/com/xamarin/android/MyCanvas.java:15: error: method save in class Canvas cannot be applied to given types; (TaskId:1188)
      return super.save (saveFlags); (TaskId:1188)
                  ^ (TaskId:1188)
    required: no arguments (TaskId:1188)
    found: int (TaskId:1188)
    reason: actual and formal argument lists differ in length (TaskId:1188)
  2 errors (TaskId:1188)
  3 warnings (TaskId:1188)
```

However, there is something *weird* here: [*There is no `Canvas.save(int)` method*](https://developer.android.com/reference/android/graphics/Canvas.html#save()).

It looks like [it was removed in API-P DP2](https://developer.android.com/sdk/api_diff/p-dp2/changes). We'll have to update the test.

**Comment by jonpryor** [comment]
build

**Comment by jonpryor** [comment]
After adding **full-mono-integration-build**, the [current build failure](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3326/):

```
  "javac" -source 1.8 -target 1.8 -d "obj/Debug/android-28/jcw/bin" -bootclasspath "/Users/builder/android-toolchain/sdk/platforms/android-P/android.jar":"/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v9.0/mono.android.jar" @obj/Debug/android-28/jcw/classes.txt (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/graphics/ImageDecoder_OnPartialImageListenerImplementor.java:28: error: cannot find symbol (TaskId:969)
        public boolean onPartialImage (android.graphics.ImageDecoder.DecodeException p0) (TaskId:969)
                                                                    ^ (TaskId:969)
    symbol:   class DecodeException (TaskId:969)
    location: class ImageDecoder (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/graphics/ImageDecoder_OnPartialImageListenerImplementor.java:33: error: cannot find symbol (TaskId:969)
        private native boolean n_onPartialImage (android.graphics.ImageDecoder.DecodeException p0); (TaskId:969)
                                                                              ^ (TaskId:969)
    symbol:   class DecodeException (TaskId:969)
    location: class ImageDecoder (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/media/MediaPlayer_OnMediaTimeDiscontinuityListenerImplementor.java:8: error: cannot find symbol (TaskId:969)
                android.media.MediaPlayer.OnMediaTimeDiscontinuityListener (TaskId:969)
                                         ^ (TaskId:969)
    symbol:   class OnMediaTimeDiscontinuityListener (TaskId:969)
    location: class MediaPlayer (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/media/MediaPlayer_OnSubtitleDataListenerImplementor.java:8: error: cannot find symbol (TaskId:969)
	                android.media.MediaPlayer.OnSubtitleDataListener (TaskId:969)
                                         ^ (TaskId:969)
    symbol:   class OnSubtitleDataListener (TaskId:969)
    location: class MediaPlayer (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/media/MediaPlayer_OnSubtitleDataListenerImplementor.java:28: error: cannot find symbol (TaskId:969)
        public void onSubtitleData (android.media.MediaPlayer p0, android.media.SubtitleData p1) (TaskId:969)
                                                                               ^ (TaskId:969)
    symbol:   class SubtitleData (TaskId:969)
    location: package android.media (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/media/MediaPlayer_OnSubtitleDataListenerImplementor.java:33: error: cannot find symbol (TaskId:969)
        private native void n_onSubtitleData (android.media.MediaPlayer p0, android.media.SubtitleData p1); (TaskId:969)
                                                                                         ^ (TaskId:969)
    symbol:   class SubtitleData (TaskId:969)
    location: package android.media (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/se/omapi/SEService_OnConnectedListenerImplementor.java:8: error: cannot find symbol (TaskId:969)
                android.se.omapi.SEService.OnConnectedListener (TaskId:969)
                                          ^ (TaskId:969)
    symbol:   class OnConnectedListener (TaskId:969)
    location: class SEService (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/view/View_OnUnhandledKeyEventListenerImplementor.java:8: error: cannot find symbol (TaskId:969)
                android.view.View.OnUnhandledKeyEventListener (TaskId:969)
                                 ^ (TaskId:969)
    symbol:   class OnUnhandledKeyEventListener (TaskId:969)
    location: class View (TaskId:969)
  obj/Debug/android-28/jcw/src/mono/android/graphics/ImageDecoder_OnPartialImageListenerImplementor.java:4: error: ImageDecoder_OnPartialImageListenerImplementor is not abstract and does not override abstract method onPartialImage(int,Source) in OnPartialImageListener (TaskId:969)
  public class ImageDecoder_OnPartialImageListenerImplementor (TaskId:969)
         ^ (TaskId:969)
```

Notably "wrong" is the path to `android.jar` used: `android-toolchain/sdk/platforms/android-P/android.jar`, which should *instead* be `android-toolchain/sdk/platforms/android-28/android.jar`.

**Pull request merged**